File: backend/app/routes/meals.py
"""Meal CRUD and offline sync.

THE CLOCK RULE, because getting this wrong fails silently:

  updated_at         client's edit clock. Used ONLY to decide which of two
                     competing edits is newer (last-write-wins).
  server_updated_at  our clock. Used ONLY as the pull cursor.

They must never be swapped. If the pull cursor came from the client's clock, a
phone running fast would push a meal stamped in the future, save that as its
cursor, and then never receive any server change written in the gap — silently,
forever. One monotonic authority for the cursor: us.
"""
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from pydantic import ValidationError
from pymongo.errors import DuplicateKeyError

from ..db import get_db
from ..schemas import MealIn, SyncRequest
from ..timeutil import floor_ms, parse_iso, to_iso, utcnow

bp = Blueprint("meals", __name__, url_prefix="/api/meals")

MAX_SYNC_BATCH = 500
TOMBSTONE_RETENTION_DAYS = 30


def _serialize(doc):
    doc = dict(doc)
    doc.pop("_id", None)
    doc.pop("user_id", None)
    for field in ("updated_at", "server_updated_at", "created_at"):
        if field in doc:
            doc[field] = to_iso(doc[field])
    return doc


def _apply_change(user_id, meal: MealIn):
    """Upsert one validated change under last-write-wins.

    Returns (status, doc) where status is 'applied' or 'conflict'. A conflict
    means the server already holds a newer edit; we hand it back so the client
    can overwrite its local copy rather than silently diverging.
    """
    client_id = meal.client_id
    edited_at = meal.updated_at or utcnow()
    meals = get_db().meals

    fields = meal.model_dump(exclude={"client_id", "updated_at"})
    fields["updated_at"] = edited_at
    fields["server_updated_at"] = utcnow()

    # Guard on updated_at so a slow retry can't clobber a newer edit. If no
    # document matched, either it doesn't exist yet or ours is stale.
    result = meals.update_one(
        {"user_id": user_id, "client_id": client_id,
         "updated_at": {"$lt": edited_at}},
        {"$set": fields},
    )
    if result.matched_count:
        return "applied", None

    try:
        meals.insert_one({
            "user_id": user_id,
            "client_id": client_id,
            "created_at": utcnow(),
            **fields,
        })
        return "applied", None
    except DuplicateKeyError:
        # It exists and its updated_at is >= ours: the server's copy wins.
        existing = meals.find_one({"user_id": user_id, "client_id": client_id})
        return "conflict", _serialize(existing)


@bp.post("/sync")
@jwt_required()
def sync():
    """Push local changes and pull everything since the last cursor."""
    try:
        payload = SyncRequest.model_validate(request.get_json(silent=True) or {})
    except ValidationError as exc:
        return jsonify({"error": "validation failed",
                        "detail": exc.errors()}), 422

    user_id = get_jwt_identity()

    # Snapshot our clock BEFORE writing, floored to Mongo's millisecond
    # storage granularity. Anything written during this request gets a
    # server_updated_at >= this value even after BSON truncation, so returning
    # it as the next cursor can never skip a change — at worst the client
    # re-receives one, which its upsert makes free. See timeutil.floor_ms.
    cursor_now = floor_ms(utcnow())

    applied, conflicts = [], []
    for meal in payload.changes:
        status, doc = _apply_change(user_id, meal)
        if status == "applied":
            applied.append(meal.client_id)
        else:
            conflicts.append(doc)

    query = {"user_id": user_id}
    since = payload.since
    if since:
        # $gte, not $gt: the boundary must overlap rather than gap.
        query["server_updated_at"] = {"$gte": floor_ms(since)}

    pulled = [
        _serialize(doc)
        for doc in get_db().meals.find(query).sort("server_updated_at", 1)
    ]

    return jsonify({
        "server_time": to_iso(cursor_now),
        "applied": applied,
        "conflicts": conflicts,
        "changes": pulled,
    })


@bp.get("")
@jwt_required()
def list_meals():
    """List meals for a day, or everything since a cursor."""
    query = {"user_id": get_jwt_identity(), "deleted": {"$ne": True}}

    day = request.args.get("date")  # YYYY-MM-DD, matched on the client's local day
    if day:
        query["eaten_at"] = {"$regex": f"^{day}"}

    since = parse_iso(request.args.get("since"))
    if since:
        query["server_updated_at"] = {"$gte": floor_ms(since)}

    docs = get_db().meals.find(query).sort("eaten_at", -1).limit(500)
    return jsonify({"meals": [_serialize(d) for d in docs]})


@bp.delete("/<client_id>")
@jwt_required()
def delete_meal(client_id):
    """Tombstone, never a hard delete — a removed row can't propagate."""
    result = get_db().meals.update_one(
        {"user_id": get_jwt_identity(), "client_id": client_id},
        {"$set": {"deleted": True, "updated_at": utcnow(),
                  "server_updated_at": utcnow()}},
    )
    if not result.matched_count:
        return jsonify({"error": "not found"}), 404
    return jsonify({"ok": True})


@bp.get("/export.csv")
@jwt_required()
def export_csv():
    """Escape hatch: your history should never be trapped in this app."""
    import csv, io

    columns = ["eaten_at", "meal_type", "name", "brand", "barcode", "serving_g",
               "calories", "protein_g", "carbs_g", "fat_g", "fiber_g",
               "sugar_g", "sodium_mg", "source", "user_edited"]

    buffer = io.StringIO()
    writer = csv.DictWriter(buffer, fieldnames=columns, extrasaction="ignore")
    writer.writeheader()
    docs = get_db().meals.find(
        {"user_id": get_jwt_identity(), "deleted": {"$ne": True}}
    ).sort("eaten_at", 1)
    for doc in docs:
        writer.writerow({c: doc.get(c, "") for c in columns})

    return buffer.getvalue(), 200, {
        "Content-Type": "text/csv; charset=utf-8",
        "Content-Disposition": 'attachment; filename="meals.csv"',
    }

File: mobile/src/db/meals.ts
import { and, desc, eq, like, sql } from 'drizzle-orm';

import type { MealPayload } from '../api/types';
import { newClientId } from '../lib/uuid';
import { db } from './client';
import { type Meal, type NewMeal, meals } from './schema';

/** Drizzle is camelCase, the wire is snake_case. Convert at exactly one place. */
export function toPayload(meal: Meal): MealPayload {
  return {
    client_id: meal.clientId,
    name: meal.name,
    brand: meal.brand,
    barcode: meal.barcode,
    serving_g: meal.servingG,
    calories: meal.calories,
    protein_g: meal.proteinG,
    carbs_g: meal.carbsG,
    fat_g: meal.fatG,
    fiber_g: meal.fiberG,
    sugar_g: meal.sugarG,
    sodium_mg: meal.sodiumMg,
    meal_type: meal.mealType,
    eaten_at: meal.eatenAt,
    source: meal.source,
    model_confidence: meal.modelConfidence,
    user_edited: meal.userEdited,
    deleted: meal.deleted,
    updated_at: meal.updatedAt,
  };
}

export function fromPayload(payload: MealPayload): NewMeal {
  return {
    clientId: payload.client_id,
    name: payload.name ?? null,
    brand: payload.brand ?? null,
    barcode: payload.barcode ?? null,
    servingG: payload.serving_g ?? null,
    calories: payload.calories ?? null,
    proteinG: payload.protein_g ?? null,
    carbsG: payload.carbs_g ?? null,
    fatG: payload.fat_g ?? null,
    fiberG: payload.fiber_g ?? null,
    sugarG: payload.sugar_g ?? null,
    sodiumMg: payload.sodium_mg ?? null,
    mealType: payload.meal_type ?? null,
    eatenAt: payload.eaten_at,
    source: payload.source,
    modelConfidence: payload.model_confidence ?? null,
    userEdited: payload.user_edited,
    deleted: payload.deleted,
    updatedAt: payload.updated_at,
    // Arrived FROM the server, so by definition it needs no push back.
    dirty: false,
  };
}

export async function createMeal(
  input: Omit<NewMeal, 'clientId' | 'updatedAt' | 'dirty'>,
): Promise<Meal> {
  const row: NewMeal = {
    ...input,
    clientId: newClientId(),
    updatedAt: new Date().toISOString(),
    dirty: true,
  };
  const [created] = await db.insert(meals).values(row).returning();
  return created;
}

export async function updateMeal(
  clientId: string,
  patch: Partial<NewMeal>,
): Promise<void> {
  await db
    .update(meals)
    .set({
      ...patch,
      // Any local edit re-stamps the edit clock and re-dirties the row, or the
      // change would never be pushed.
      updatedAt: new Date().toISOString(),
      dirty: true,
    })
    .where(eq(meals.clientId, clientId));
}

/** Tombstone, never a hard delete — a removed row can't propagate to the server. */
export async function deleteMeal(clientId: string): Promise<void> {
  await updateMeal(clientId, { deleted: true });
}

/** `day` is YYYY-MM-DD in the device's local time. */
export async function listMealsForDay(day: string): Promise<Meal[]> {
  return db
    .select()
    .from(meals)
    .where(and(eq(meals.deleted, false), like(meals.eatenAt, `${day}%`)))
    .orderBy(desc(meals.eatenAt));
}

export async function getMeal(clientId: string): Promise<Meal | undefined> {
  const [row] = await db
    .select()
    .from(meals)
    .where(eq(meals.clientId, clientId))
    .limit(1);
  return row;
}

/** Everything with unpushed edits — the push half of a sync. */
export async function dirtyMeals(): Promise<Meal[]> {
  return db.select().from(meals).where(eq(meals.dirty, true));
}

export async function markClean(clientIds: string[]): Promise<void> {
  if (clientIds.length === 0) return;
  await db
    .update(meals)
    .set({ dirty: false })
    .where(sql`${meals.clientId} IN ${clientIds}`);
}

/**
 * Apply a server-authored row, overwriting whatever is local.
 *
 * Used for both pulled changes and conflict resolutions. Safe to call twice
 * with the same row — which is exactly why the sync cursor is allowed to
 * overlap rather than risk a gap.
 */
export async function upsertFromServer(payload: MealPayload): Promise<void> {
  const row = fromPayload(payload);
  await db
    .insert(meals)
    .values(row)
    .onConflictDoUpdate({ target: meals.clientId, set: row });
}

/** Frequently-eaten foods, for one-tap re-logging. Phase 0.5. */
export async function recentDistinctMeals(limit = 20) {
  return db
    .select({
      name: meals.name,
      brand: meals.brand,
      barcode: meals.barcode,
      servingG: meals.servingG,
      calories: meals.calories,
      proteinG: meals.proteinG,
      carbsG: meals.carbsG,
      fatG: meals.fatG,
      uses: sql<number>`count(*)`.as('uses'),
      lastEaten: sql<string>`max(${meals.eatenAt})`.as('last_eaten'),
    })
    .from(meals)
    .where(and(eq(meals.deleted, false), sql`${meals.name} is not null`))
    .groupBy(meals.name, meals.brand)
    .orderBy(desc(sql`uses`), desc(sql`last_eaten`))
    .limit(limit);
}
